util/mon: reserve extra bytes for allocation in BytesAccount

petermattis opened the pull request:
```
name                time/op
BoundAccountGrow-8  61.6ns ± 5%
```

This is very expensive. For comparison, pushing a row through a distsql
`RowChannel` takes ~100ns (which is also very expensive).

Release note: None

petermattis committed:
util/mon: add BenchmarkBoundAccountGrow

name                time/op
BoundAccountGrow-8  61.6ns ± 5%

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/20590)
<!-- Reviewable:end -->

knz commented:
I know! 

1) Originally when I wrote this code in #9259 I was very adamant that only the connection between one monitor and another should use a mutex (via a separate abstraction `MemoryPool`) but not the connection between an account and a monitor. The original design was saying "a monitor is going to allocate in chunks from another monitor, so that the accounts that bill against it can use more fine grained grow operations without the mutex overhead". You can check the commit history in #9259 to verify this is true. Then Andrei came after me and said "this is too complicated, we want accounts to bill against monitors concurrently, so put a mutex in the monitor already and drop the pool". I knew this was bad but at that point I was tired from arguing.

2) I suspect the `if log.V(2)` in there is also not helping. I am not sure how to make this better -- we (at least, I) have found the conditional logging useful in the past, but `log.V` is not exactly cheap.

Feel free to find someone to think again about the original design and do the arguing.

petermattis committed:
util/mon: reserve extra bytes for allocation in BytesAccount

Previously, BytesAccount was always maintaining exactly the bytes
requested by the user. This meant that increasing the allocation for a
BytesAccount need to increase the allocation on the associated monitor
which is somewhat expensive. Added BytesAccount.reserved so that
allocations are served from this local cache of accounted for bytes
first. This avoids expensive locking in the common case.

name                old time/op  new time/op  delta
BoundAccountGrow-8  60.7ns ± 3%   5.8ns ± 2%  -90.40%  (p=0.000 n=10+10)

Fixes #20591

Release note (performance improvement): Significantly reduce the
overhead of SQL memory accounting.

petermattis commented:
I added a commit which fixes the performance of `BoundAccount.Grow` by adding `BytesAccount.reserved` and allocating in `poolAllocationSize` chunks when growing a `BoundAccount`. This was much easier than trying to do anything with `BytesMonitor` and gets us near ideal performance for allocation tracking (a few nanoseconds per-op).

Cc @andreimatei, @RaduBerinde as @knz might be offline this week.

asubiotto commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 3 files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20590#-:-L052NlcTNgfHJlCAXBo:bnfp4nl)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
LGTM

Please change the PR description from the 1st commit to the 2nd.

---

Review status: 0 of 3 files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[pkg/util/mon/bytes_usage.go, line 421 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20590#-L05JPP4v0Q--did3dw8:-L05JPP4v0Q--did3dw9:b-x6tmv5) ([raw file](https://github.com/cockroachdb/cockroach/blob/756d379eb1904022d21f262dc9fa3e392bbd2c63/pkg/util/mon/bytes_usage.go#L421)):*
> ```Go
> type BytesAccount struct {
> 	curAllocated int64
> 	reserved     int64
> ```

comment pls. Should say that reserved drops as curAllocated increases (i.e. it is not static).

---

*[pkg/util/mon/bytes_usage.go, line 426 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20590#-L05MymPYWOev3mpp1ow:-L05MymPYWOev3mpp1ox:bvwbsry) ([raw file](https://github.com/cockroachdb/cockroach/blob/756d379eb1904022d21f262dc9fa3e392bbd2c63/pkg/util/mon/bytes_usage.go#L426)):*
> ```Go
> // CurrentlyAllocated returns the number of bytes currently allocated through
> // this account.
> func (acc BytesAccount) CurrentlyAllocated() int64 {
> ```

nit: to me `CurrentlyAllocated` sounds like it should return what `totalAllocated` returns.
Unfortunately, it appears that we also need a method that returns `curAllocated` for the `hashJoiner`. I think a better name for that might be `Filled()`.

---

*[pkg/util/mon/bytes_usage.go, line 441 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20590#-L05LcFz2JKB-_jYpbGJ:-L05LcFz2JKB-_jYpbGK:b14ct2y) ([raw file](https://github.com/cockroachdb/cockroach/blob/756d379eb1904022d21f262dc9fa3e392bbd2c63/pkg/util/mon/bytes_usage.go#L441)):*
> ```Go
> // OpenAndInitAccount creates a new account and pre-allocates some
> // initial amount of bytes.
> func (mm *BytesMonitor) OpenAndInitAccount(
> ```

please comment something here about how the new account is going to look. It's going to have `curAllocated = initialAllocation` and `reserved` set to  some rounding, right?

Generally, this rounding is the a big deal because that's why this PR achieves anything, right? I think that should be impressed in a comment on `BoundAccount`.

---

*[pkg/util/mon/bytes_usage.go, line 450 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20590#-L05KVt4DHEvDfQJMnhO:-L05KVt4DHEvDfQJMnhP:b-bxw8a4) ([raw file](https://github.com/cockroachdb/cockroach/blob/756d379eb1904022d21f262dc9fa3e392bbd2c63/pkg/util/mon/bytes_usage.go#L450)):*
> ```Go
> // GrowAccount requests a new allocation in an account.
> func (mm *BytesMonitor) GrowAccount(ctx context.Context, acc *BytesAccount, extraSize int64) error {
> 	if acc.reserved < extraSize {
> ```

I think it'd be better if we kept the `BytesMonitor` unaware of `BoundAccount.reserved`. Like, handle the case where `extraSize < acc.reserved` in `BoundAccount` directly.

---

*[pkg/util/mon/bytes_usage.go, line 572 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20590#-L05Jy664_AGwbnyDxb6:-L05Jy664_AGwbnyDxb7:b-1za4sk) ([raw file](https://github.com/cockroachdb/cockroach/blob/756d379eb1904022d21f262dc9fa3e392bbd2c63/pkg/util/mon/bytes_usage.go#L572)):*
> ```Go
> 	// Check the local limit first. NB: The condition is written in this manner
> 	// so that it handles overflow correctly. Consider what happens if
> 	// x==math.MaxInt64. mm.limit-x will be a large negative number.
> ```

you're making a big deal out of this, but do we ever reserver MaxInt64 bytes? What does that mean?
Cause if we don't, I'd revert this change.

---

*[pkg/util/mon/bytes_usage_test.go, line 326 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20590#-L05M2YyoI1GdIHNxq-w:-L05M2YyoI1GdIHNxq-x:b7p4xwg) ([raw file](https://github.com/cockroachdb/cockroach/blob/756d379eb1904022d21f262dc9fa3e392bbd2c63/pkg/util/mon/bytes_usage_test.go#L326)):*
> ```Go
> 
> 	ctx := context.Background()
> 	m := MakeMonitor("test", MemoryResource, nil, nil, 1e9, 1e9)
> ```

lotsa nils and literals :)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20590#-:-L05IlvAU3PQgpBDoh9D:bb8ibj8)*
<!-- Sent from Reviewable.io -->

petermattis commented:
PR title updated, though does that matter? The commit titles are what we usually look at historically.

---

Review status: 0 of 3 files reviewed at latest revision, 6 unresolved discussions.

---

*[pkg/util/mon/bytes_usage.go, line 421 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20590#-L05JPP4v0Q--did3dw8:-L05Y11KfJD2MI_wQQei:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/756d379eb1904022d21f262dc9fa3e392bbd2c63/pkg/util/mon/bytes_usage.go#L421)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

comment pls. Should say that reserved drops as curAllocated increases (i.e. it is not static).
</blockquote></details>

Done.

---

*[pkg/util/mon/bytes_usage.go, line 426 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20590#-L05MymPYWOev3mpp1ow:-L05Y4qgiVeNHOnBOP8c:budy3xh) ([raw file](https://github.com/cockroachdb/cockroach/blob/756d379eb1904022d21f262dc9fa3e392bbd2c63/pkg/util/mon/bytes_usage.go#L426)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

nit: to me `CurrentlyAllocated` sounds like it should return what `totalAllocated` returns.
Unfortunately, it appears that we also need a method that returns `curAllocated` for the `hashJoiner`. I think a better name for that might be `Filled()`.
</blockquote></details>

`CurrentlyAllocated` makes sense to me given what it returns. Are you suggesting a different name for `totalAllocated()`?

---

*[pkg/util/mon/bytes_usage.go, line 441 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20590#-L05LcFz2JKB-_jYpbGJ:-L05YWCm_x4vP4wxOcpM:b2cok8l) ([raw file](https://github.com/cockroachdb/cockroach/blob/756d379eb1904022d21f262dc9fa3e392bbd2c63/pkg/util/mon/bytes_usage.go#L441)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

please comment something here about how the new account is going to look. It's going to have `curAllocated = initialAllocation` and `reserved` set to  some rounding, right?

Generally, this rounding is the a big deal because that's why this PR achieves anything, right? I think that should be impressed in a comment on `BoundAccount`.
</blockquote></details>

Done and done (extended the comment on `BytesAccount`)

---

*[pkg/util/mon/bytes_usage.go, line 450 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20590#-L05KVt4DHEvDfQJMnhO:-L05Yy0rpownhOnuzxJh:b-oxkxrc) ([raw file](https://github.com/cockroachdb/cockroach/blob/756d379eb1904022d21f262dc9fa3e392bbd2c63/pkg/util/mon/bytes_usage.go#L450)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I think it'd be better if we kept the `BytesMonitor` unaware of `BoundAccount.reserved`. Like, handle the case where `extraSize < acc.reserved` in `BoundAccount` directly.
</blockquote></details>

You're getting confused between `BytesAccount` and `BoundAccount`. `BoundAccount.reserved` does not exist. We could move this code to `BoundAccount.Grow`, but I'm not at all clear that doing so is better given that there are some direct calls to `BytesMonitor.GrowAccount` that would then miss out on the action.

---

*[pkg/util/mon/bytes_usage.go, line 572 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20590#-L05Jy664_AGwbnyDxb6:-L05ZPmNaWpzngoqS_KD:b-p3hd63) ([raw file](https://github.com/cockroachdb/cockroach/blob/756d379eb1904022d21f262dc9fa3e392bbd2c63/pkg/util/mon/bytes_usage.go#L572)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

you're making a big deal out of this, but do we ever reserver MaxInt64 bytes? What does that mean?
Cause if we don't, I'd revert this change.
</blockquote></details>

We have `SELECT repeat('s', 9223372036854775807)` in the SQL logic tests. This was accidentally getting rejected before. Now we have a more principled rejection. 

---

*[pkg/util/mon/bytes_usage_test.go, line 326 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20590#-L05M2YyoI1GdIHNxq-w:-L05ZmQUq2ztDQx2dDRe:blkxlnx) ([raw file](https://github.com/cockroachdb/cockroach/blob/756d379eb1904022d21f262dc9fa3e392bbd2c63/pkg/util/mon/bytes_usage_test.go#L326)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

lotsa nils and literals :)
</blockquote></details>

inline comments added. I was just copying from other code in this file.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20590#-:-L05_3X4gsBlkW0uGnz0:b-xm1b3y)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
> PR title updated, though does that matter? The commit titles are what we usually look at historically.

Sometimes people search for PRs on Github, especially when they're really interested in a comment or the review activity more than the code changes.

tschottdorf commented:
(besides, the PR title leaks into the merge commit)

petermattis commented:
Review status: 0 of 3 files reviewed at latest revision, 6 unresolved discussions.

---

*[pkg/util/mon/bytes_usage.go, line 450 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20590#-L05KVt4DHEvDfQJMnhO:-L05aj71pBZaC9Ijofyy:b-3634qk) ([raw file](https://github.com/cockroachdb/cockroach/blob/756d379eb1904022d21f262dc9fa3e392bbd2c63/pkg/util/mon/bytes_usage.go#L450)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

You're getting confused between `BytesAccount` and `BoundAccount`. `BoundAccount.reserved` does not exist. We could move this code to `BoundAccount.Grow`, but I'm not at all clear that doing so is better given that there are some direct calls to `BytesMonitor.GrowAccount` that would then miss out on the action.
</blockquote></details>

Also, `BytesMonitor.GrowAccount` nows about `BytesAccount.curAllocated`. Why wouldn't it know about `BytesAccount.reserved`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20590)*
<!-- Sent from Reviewable.io -->

petermattis commented:
@tschottdorf TIL

---

Review status: 0 of 3 files reviewed at latest revision, 6 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20590#-:-L05ax2lstHIhsGaYOlC:bcktfza)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: 0 of 3 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[pkg/util/mon/bytes_usage.go, line 426 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20590#-L05MymPYWOev3mpp1ow:-L05jBfXLIfD1Wnf9aNC:buesq27) ([raw file](https://github.com/cockroachdb/cockroach/blob/756d379eb1904022d21f262dc9fa3e392bbd2c63/pkg/util/mon/bytes_usage.go#L426)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

`CurrentlyAllocated` makes sense to me given what it returns. Are you suggesting a different name for `totalAllocated()`?
</blockquote></details>

This is all nits.
`CurrentlyAllocated` is the name that bothers me the most. The "allocated" memory is the sum. The fact that it's not all in use does not make the memory any less "allocated".

---

*[pkg/util/mon/bytes_usage.go, line 450 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20590#-L05KVt4DHEvDfQJMnhO:-L05kGEFDxBM23jvpPHs:bi3abk0) ([raw file](https://github.com/cockroachdb/cockroach/blob/756d379eb1904022d21f262dc9fa3e392bbd2c63/pkg/util/mon/bytes_usage.go#L450)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Also, `BytesMonitor.GrowAccount` nows about `BytesAccount.curAllocated`. Why wouldn't it know about `BytesAccount.reserved`?
</blockquote></details>

In my ideal world, `BytesAccount` would go away and we'd only have `BoundAccount` (which could then be named `Account`). I never understood the need for the non-bound one.
Moving the code to `BoundAccount.Grow` is indeed what I was thinking of.
Anyway, whatevs you think is best.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20590)*
<!-- Sent from Reviewable.io -->

petermattis committed:
util/mon: minor renamings

BytesAccount.curAllocated       -> BytesAccount.used
BytesAccount.CurrentlyAllocated -> BytesAccount.Used
BytesAccount.totalAllocated     -> BytesAccount.allocated

Release note: None

petermattis committed:
sql,util/mon: remove external usage of mon.BytesAccount

Unexport mon.BytesAccount. Replace existing usage with
mon.BoundAccount. Unexport various methods on BytesMonitor. BoundAccount
is the interface to use. This is preparation for merging BytesAccount
and BoundAccount.

Release note: None

petermattis committed:
util/mon: merge BytesAccount and BoundAccount

Also inline the various BytesMonitor.*Account() methods directly into
their sole callsites in BoundAccount. This simplifies the implementation
as well as provides a decent performance boost:

name                old time/op  new time/op  delta
BoundAccountGrow-8  5.62ns ± 1%  2.87ns ± 2%  -49.04%  (p=0.000 n=10+8)

Release note: None

petermattis committed:
util/mon: rename BoundAccount.ResizeItem to BoundAccount.Resize

For consistency with the other BoundAccount methods, none of which
mentions "items".

Release note: None

petermattis commented:
PTAL. Can you scrutinize the changes to `prepared_stmt.go`? It wasn't clear to me what was going on there and why a `BoundAccount` wasn't being used.

---

Review status: 0 of 3 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[pkg/util/mon/bytes_usage.go, line 426 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20590#-L05MymPYWOev3mpp1ow:-L05rmAWoEJaYI2EqHLM:b90vv4r) ([raw file](https://github.com/cockroachdb/cockroach/blob/756d379eb1904022d21f262dc9fa3e392bbd2c63/pkg/util/mon/bytes_usage.go#L426)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

This is all nits.
`CurrentlyAllocated` is the name that bothers me the most. The "allocated" memory is the sum. The fact that it's not all in use does not make the memory any less "allocated".
</blockquote></details>

Did some renamings. PTAL.

---

*[pkg/util/mon/bytes_usage.go, line 450 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20590#-L05KVt4DHEvDfQJMnhO:-L06DgujSEacN6JJYyJf:bmromgy) ([raw file](https://github.com/cockroachdb/cockroach/blob/756d379eb1904022d21f262dc9fa3e392bbd2c63/pkg/util/mon/bytes_usage.go#L450)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

In my ideal world, `BytesAccount` would go away and we'd only have `BoundAccount` (which could then be named `Account`). I never understood the need for the non-bound one.
Moving the code to `BoundAccount.Grow` is indeed what I was thinking of.
Anyway, whatevs you think is best.
</blockquote></details>

I wasn't intending do this, but I discovered that moving this work to `BoundAccount.Grow` provided a decent performance boost. 

I've added a few more commits which simplify everything. `BytesAccount` is now gone. We just have `BytesMonitor` and `BoundAccount`. I suppose I could rename those as well, but go-rename isn't working for me right now.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20590#-:-L06EHSA54FdS_a3TIJ2:br4pnhn)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: 0 of 8 files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[pkg/sql/executor.go, line 488 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/20590#-L06Jx_ecgNGkl6nY7S4:-L06Jx_ecgNGkl6nY7S5:bwtyep7) ([raw file](https://github.com/cockroachdb/cockroach/blob/63733938bf5f93d355f9a5b3b0127aada74c26c2/pkg/sql/executor.go#L488)):*
> ```Go
> 		TypeHints:   placeholderHints,
> 		portalNames: make(map[string]struct{}),
> 		// We need a memory account available in order to prepare a statement, since we
> ```

I'd remove this comment now. It talks about constant folding, although that no longer matches the lifetime of this `memAcc` (it didn't really match it before either). And if anything, the comment should be on the struct.

I also have failed to really understand what was going on with `constantAcc/memAcc`. Perhaps this is an indication that some more pulling would be good :). My impression is that the intention was to separate allocations that are long lived (currently, this line `sz := int64(uintptr(len(name)+len(stmtStr)) + unsafe.Sizeof(*pStmt))`)  from allocations performed during planning the prepared statement (this business with the constant folding, I dunno)
As we were talking, let's make a temp acct before the call to `planner.Prepare()` and close it immediately.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20590)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 8 files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20590#-:-L06NElu_Cd041LfkTkR:bnfp4nl)*
<!-- Sent from Reviewable.io -->

petermattis committed:
sql: tweak memory accounting for prepared statements

Clear the memory account used for planning a prepared statement when we
close the prepared statement.

Release note: None

petermattis commented:
Review status: 0 of 8 files reviewed at latest revision, 6 unresolved discussions.

---

*[pkg/sql/executor.go, line 488 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/20590#-L06Jx_ecgNGkl6nY7S4:-L06NRTveIIbL-ftRNJA:b-rd90mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/63733938bf5f93d355f9a5b3b0127aada74c26c2/pkg/sql/executor.go#L488)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I'd remove this comment now. It talks about constant folding, although that no longer matches the lifetime of this `memAcc` (it didn't really match it before either). And if anything, the comment should be on the struct.

I also have failed to really understand what was going on with `constantAcc/memAcc`. Perhaps this is an indication that some more pulling would be good :). My impression is that the intention was to separate allocations that are long lived (currently, this line `sz := int64(uintptr(len(name)+len(stmtStr)) + unsafe.Sizeof(*pStmt))`)  from allocations performed during planning the prepared statement (this business with the constant folding, I dunno)
As we were talking, let's make a temp acct before the call to `planner.Prepare()` and close it immediately.
</blockquote></details>

Removed the comment. Rather than create another account for planning, I now I call `preparedStmt.memAcc.Clear()` when we close the plan node. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20590)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 8 files reviewed at latest revision, 6 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20590#-:-L06RG0c0c-pxCHDePsB:bnfp4nl)*
<!-- Sent from Reviewable.io -->

petermattis committed:
sql: prepared statement memory account should be tied to Session.sessionMon

The difference between Session.mon and Session.sessionMon is confusing.

Release note: None
